csrc/fusion.cpp

```diff
@@ -172,14 +172,8 @@ void Fusion::removeExpr(Expr* expr) {
     out->setDefinition(nullptr);
   }
 
-  for (auto inp : expr->inputs()) {
-    auto uses_copy = inp->uses();
-    auto it = std::find(uses_copy.begin(), uses_copy.end(), expr);
-    if (it != uses_copy.end()) {
-      uses_copy.erase(it);
-      inp->setUses(uses_copy);
-    }
-  }
+  // Call for a rebuild of uses vector
+  invalidateUses();
```

Comment:
I think I said we should just do this when removing an expr, but that's only for tensors because `resetTvUses` doesn't reset uses of scalars. Sorry for not very clear.

Reply:
If an Expr is removed, it should no longer appear in any of its inputs' `uses`, even for Scalars right? So we could either remove `expr` from those manually or rebuild them all. Note that on `main` this line is just calling `resetTvUses()`; does that mean Scalar uses are not updated currently in removeExpr?

Reply:
> If an Expr is removed, it should no longer appear in any of its inputs' uses, even for Scalars right?

Yes

> rebuild them all which is not done by resetTvUses()

Yes, that's the source of the confusion. `resetTvUses` only resets uses of tensors. So for scalars, we need to remove it from the use lists of the inputs.

> Note that on main this line is just calling resetTvUses();

Is it true? `resetTvUses` is called through `inp->uses()`, but there's actual remove happening after that at lines around 177-182.

Reply:
You're right. And when I removed the actual "removal" it was causing a segfault in `FusionGridReduction3dim0_CUDA`. In latest commit I added a `removeUse` to complement `addUse` which will call `invalidateTvUses` for TV inputs, and for non-TVs it will do the actual removal from `uses_`.